docs_src/design/adr/014-Secret-Provider-For-All.md

```diff
@@ -0,0 +1,409 @@
+    + [Mocks](#mocks)
+    + [Where will `SecretProvider` reside?](#where-will--secretprovider--reside-)
+      - [Go Services](#go-services)
+      - [C Device Service](#c-device-service)
+  * [Consequences](#consequences)
+
+## Status
+
+**proposed**
+
+## Context
+
+This ADR defines the new `SecretProvider` abstraction that will be used by all EdgeX services, including Device Services. The Secret Provider is used by services to retrieve secrets from the Secret Store. The Secret Store, in secure mode, is currently Vault. In non-secure mode it is configuration in some form, i.e. `DatabaseInfo` configuration or `InsecureSecrets` configuration for Application Services.
+
+### Existing Implementations
+
+The Secret Provider abstraction defined in this ADR is based on the Secret Provider abstraction implementations in the Application Functions SDK (App SDK) for Application Services and the one in go-mod-bootstrap (Bootstrap) used by the Core, Support & Security services in edgex-go. Device Services do not currently use secure secrets. The App SDK implementation was initially based on the Bootstrap implementation.
+
+The similarities and differences between these implementations are:
+
+- Both wrap the `SecretClient` from go-mod-secrets
+- Both initialize the `SecretClient` based on the `SecretStore` configuration(s)
+- Both have factory method, but they differ greatly
+- Both implement the `GetDatabaseCredentials` API
+- Bootstrap's uses split interfaces definitions (`CredentialsProvider` & `CertificateProvider`) while the App SDK's use a single interface (`SecretProvider`) for the abstraction 
+- Bootstrap's includes the bootstrap handler while the App SDK's has the bootstrap handler separated out
+- Bootstrap's implements the `GetCertificateKeyPair` API, which the App SDK's does not
+- App SDK's implements the following, which the Bootstrap's does not
+  - `Initialize` API (Bootstrap's initialization is done by the bootstrap handler)
+  - `StoreSecrets` API 
+  - `GetSecrets` API
+  - `InsecureSecretsUpdated` API
+  - `SecretsLastUpdated` API
+  - Wraps a second `SecretClient` for the Application Service instance's exclusive secrets.
+    - Used by the `StoreSecrets` & `GetSecrets` APIs
+  - The standard `SecretClient` is considered the shared client for secrets that all Application Service instances share. It is only used by the `GetDatabaseCredentials` API
+  - Configuration based secret store for non-secure mode called `InsecureSecrets`
+  - Caching of secrets
+    - Needed so that secrets used by pipeline functions do not cause call out to Vault for every Event processed
+
+#### What is a Secret?
+
+A secret is a collection of key/value pairs stored in a `SecretStore` at specified path whose values are sensitive in nature. Redis database credentials are an example of a `Secret` which contains the `username` and `password` key/values stored at the `redisdb` path.
```

Comment:
This is probably out-of-scope for this ADR, but one thing that's not super clear to me is the fact that we have per service credentials (username/password) for Redis, but we also have a single password for Redis which is read by security-bootstrap-redis, which in turn sets this as the Redis password. I'm not quite sure how these credentials relate?

Reply:
Yes, out of scope for this ADR, but that is a Redis thing in that it didn't support the creation of a username as other DBs. I saw a not that Redis 6 does. So we should look at using that feature moving forward.

Reply:
> Yes, out of scope for this ADR, but that is a Redis thing in that it didn't support the creation of a username as other DBs. I saw a note that Redis 6 does. So we should look at using that feature moving forward.

I read the [docs](https://redis.io/topics/acl) and I am not convinced the new ACL system supports persistent usernames and passwords.  It still seems to be dynamic or relying on either redis.conf or an external ACL file.